### PR TITLE
[RubyGemsGemInstaller] Validate checksums from the compact index

### LINKED ISSUE #4464: Use the gem hash to throw an exception on mismatch
From here https://github.com/bundler/bundler/issues/4461#issuecomment-214003067

The index format includes a hash of the gem contents, bundler can check those vs local to see if the content is mismatched

### PR DESCRIPTION
**Pull request by segiddins**
Closes #4464

### CODE REVIEW

**Comment by indirect** [comment]
woo @homu r+

**Comment by indirect** [comment]
Ugh, of course now that I've approved it I realized that we should add a setting to disable this check, because some people will have existing gems with invalid checksums that they want to keep using.

**Comment by indirect** [comment]
@homu r-

**Comment by indirect** [comment]
I am super happy with this, though!

File: spec/support/artifice/compact_index.rb [hunk 1]
```diff
@@ -78,7 +78,12 @@ def gems(gem_repo = GEM_REPO)
               reqs = d.requirement.requirements.map {|r| r.join(" ") }.join(", ")
               CompactIndex::Dependency.new(d.name, reqs)
             end
-            CompactIndex::GemVersion.new(spec.version.version, spec.platform.to_s, nil, nil,
+            checksum = begin
+                         Digest::SHA256.file("#{GEM_REPO}/gems/#{spec.original_name}.gem").base64digest
+                       rescue
+                         nil
+                       end
+            CompactIndex::GemVersion.new(spec.version.version, spec.platform.to_s, checksum, nil,
```
Comment (indirect) [inline spec/support/artifice/compact_index.rb]: can we just `checksum = Digest::SHA256.file("#{GEM_REPO}/gems/#{spec.original_name}.gem").base64digest rescue nil`? lots of lines super far indented makes me feel gross. 😝

Comment (segiddins) [inline spec/support/artifice/compact_index.rb]: I don't like the modifier `rescue` and rubocop agrees with me :P

**Comment by segiddins** [comment]
@indirect added a setting to disable validation

**Comment by segiddins** [comment]
This is failing because `index.rubygems.org` is inconsistent between hex & base64 encoding the checksums. Is it worth accounting for that or should we convert to one or the other server-side?

File: spec/support/artifice/compact_index.rb (hunk 1 above)
Comment (indirect) [inline spec/support/artifice/compact_index.rb]: how about `checksum = gem_checksum(spec)` then?

File: lib/bundler/rubygems_gem_installer.rb [hunk 2]
```diff
@@ -12,5 +12,33 @@ def self.at(*args)
...
+        io.rewind
+        digest.base64digest!
+      end
+      unless digest == checksum
+        raise SecurityError,
+          "The checksum for the downloaded `#{spec.full_name}.gem` did not match " \
+          "the checksum given by the API. This means that the contents of the " \
+          "gem appear to be different from what was uploaded, and could be an indicator of a security issue.\n" \
+          "(The expected SHA256 checksum was #{checksum.inspect}, but the checksum for the downloaded gem was #{digest.inspect}.)\n" \
+          "Bundler cannot continue installing #{spec.name} (#{spec.version})."
```
Comment (indirect) [inline lib/bundler/rubygems_gem_installer.rb]: Based on the principle of empowering users to resolve their own errors, how do you feel about also printing the full path, suggesting deleting the gem with the bad checksum, and explaining that if they are sure they want to install this gem despite the checksum not matching, they can `bundle config disable.checksum_validaiton true`?

File: lib/bundler/rubygems_gem_installer.rb [hunk 3]
```diff
@@ -12,5 +12,33 @@ def self.at(*args)
...
     end
+
+    def pre_install_checks
+      super && validate_bundler_checksum(options[:bundler_expected_checksum])
+    end
+
+  private
+
+    def validate_bundler_checksum(checksum)
+      return true if Bundler.settings[:disable_checksum_validation]
```
Comment (indirect) [inline lib/bundler/rubygems_gem_installer.rb]: Can we make the setting `disable.checksum_validation`?

**Comment by indirect** [comment]
The config key should probably also to be added to the `bundle config` manpage section about available config keys.

**Comment by indirect** [comment]
Running a data migration right now to replace all checksums in the bundler-api database with the longer hex versions. I'm torn about what we should put in the actual index files, though... using base64 checksums would decrease the file size. :/

File: lib/bundler/rubygems_gem_installer.rb (hunk 3 above)
Comment (segiddins) [inline lib/bundler/rubygems_gem_installer.rb]: none of the other bool keys have dots in them, and wouldn't that break doing the setting as an env var?

**Comment by segiddins** [comment]
Rebased

**Comment by indirect** [comment]
This should be good now that the checksums in the database are all consistent.

@homu r+

**Comment by segiddins** [comment]
This should pass now

**Comment by indirect** [comment]
@homu retry

**Comment by indirect** [comment]
oops, @homu r+

**Pull request merged**